### PR TITLE
LSP - show references in package

### PR DESCRIPTION
**Pull request by aadi-stripe**
Adds functionality to LSP references method to show references within a package.

```
                // Case 1. get-refs on a package declaration
                //   class Foo < PackageSpec
                //         ^^^
                //   Returns all `import Foo` statements, globally
                //
                // Case 2. get-refs on an import statement
                //
                //  class Foo < PackageSpec
                //    ...
                //
                //    import Bar
                //          ^^^
                //  Returns all usages of `Bar` or `Test::Bar` *within* the Foo package only.
                //
                // Case 3. get-refs on an export statement
                //
                //  class Foo < PackageSpec
                //    ...
                //
                //    export Foo::A
                //                ^
                //  Returns all global usages of Foo::A
```

### Motivation

https://github.com/sorbet/sorbet/issues/6202

### Test plan

See included automated tests. Tested in Stripe codebase.

### CODE REVIEW

**Comment by aadi-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_MH7Z5tSj5dDLDe
→ https://go/builds/bui_MH7ZbZAX4vQHHc

**Review by froydnj — COMMENTED** [review]
I think a proper code review of this will have to wait until next week, but at first glance it seems fairly reasonable.

I think the high-order bit about the design is that, AIUI, #6202 is proposing custom Sorbet LSP requests (e.g. `sorbet/showSymbol`) to implement this functionality, whereas this PR makes `textDocument/references` work slightly differently on `__package.rb` files.

The design here is convenient, but it does mean that we are leaking implementation details of Stripe packages into LSP when `--stripe-packages` is not active.  I'm not sure whether this is the right thing to do.  It also changes the behavior of find all references, which might be surprising to some people -- though it's also more discoverable this way than putting it behind a context menu.

I also think it's worth thinking about whether we could make this work if your cursor was on the `import` token itself rather than just the constant.

**Comment by aadi-stripe** [comment]
> I think a proper code review of this will have to wait until next week, but at first glance it seems fairly reasonable.
> 
> I think the high-order bit about the design is that, AIUI, #6202 is proposing custom Sorbet LSP requests (e.g. `sorbet/showSymbol`) to implement this functionality, whereas this PR makes `textDocument/references` work slightly differently on `__package.rb` files.
> 
> The design here is convenient, but it does mean that we are leaking implementation details of Stripe packages into LSP when `--stripe-packages` is not active. I'm not sure whether this is the right thing to do. It also changes the behavior of find all references, which might be surprising to some people -- though it's also more discoverable this way than putting it behind a context menu.
> 
> I also think it's worth thinking about whether we could make this work if your cursor was on the `import` token itself rather than just the constant.

Thanks for the comments. A few thoughts on custom method vs piggybacking on `references` -- fwiw it was an option I aligned on with @jez in the RP channel in Slack.

1. The old `--stripe-packages` design had generally this same behavior, and hence did also leak implementation details into LSP. Personally since `__package.rb` files are special and purely a feature of `--stripe-packages`, I don't think it's too bad for LSP to work differently in those files.
2. The existing get-refs behavior in package files is actually not very meaningful or useful. On doing a get-ref on imports, it returns other import statements of the constant in other packages.
3. As you pointed out, a custom method in this case would be less ergonomic to users, since the existing behavior is not very useful, and to get the useful behavior they would have to call a different command.
4. We could document the difference in behavior so that people are aware that package files are treated differently by design.

Agree that it would be worth doing this on `import` statements. But personally I think that that should be an additional feature to this, not in lieu of.

**Comment by froydnj** [comment]
Thanks for the context!  Reassigning since it sounds like jez has more context here than I do.

File: main/lsp/LSPTask.cc [hunk 1]
```diff
@@ -54,7 +54,7 @@ ResponseMessageStatus statusForResponse(const ResponseMessage &response) {
                     // textDocument/documentSymbol
                     return ResponseMessageStatus::Unknown;
                 } else if constexpr (is_same_v<T, variant<JSONNullObject, vector<unique_ptr<Location>>>>) {
-                    // textDocument/definition, textDocument/typeDefinition, textDocument/references, and
+                    // textDocument/definition, textDocument/typeDefinition, textDocument/references and
```
Comment (jez) [inline main/lsp/LSPTask.cc:57]: You don't like Oxford commas?

Comment (aadi-stripe) [reply main/lsp/LSPTask.cc:57]: Complete accident, but come to think of it I'm not a fan lol. 

Will revert.

**Comment by aadi-stripe** [comment]
@jez ptal

File: README.md [hunk 2]
```diff
@@ -480,6 +480,57 @@ different version number, and also marked `default-arg-value`:
...
 initialize the variable from the argument passed at the send, or to the default value when no value is present.
 
+Finding all references works differently in package specification (__package.rb) files. Consider the following:
+
+```ruby
+class Foo < PackageSpec
+  import Bar
+```
+
+Calling "find all references" on `Bar` in this file will return only references to `Bar` in the `Foo` package. LSP tests
```
Comment (jez) [inline README.md:490]: What does this mean Go to Definition will report for `import Bar` here?

Comment (aadi-stripe) [reply README.md:490]: Go to definition on `import Bar` will go to the declaration of the `Bar` package -- that behavior doesn't change.

File: README.md [hunk 3]
```diff
@@ -480,6 +480,57 @@ different version number, and also marked `default-arg-value`:
...
+  class Foo::Baz
+    Bar.new
+ #  ^^^ importusage: bar
+  end
+```
+
+With these annotations, the LSP test will check if "find all references" on `Bar` in `import Bar` statement returns the `Bar.new` usage.
+
+Note that an `import` assertion is dissimilar to a `def` assertion, in that it is in fact a subclass of a `usage` assertion. 
+In this case, the `def` corresponding to an `import` is the PackageSpec declaration of the imported package. Calling "find all references"
```
Comment (jez) [inline README.md:509]: So upon arriving at the `class Bar < PackageSpec` location, clicking on go-to-def for the `class Bar` takes you right there, is that correct?

IIRC, `class Bar < PackageSpec` is secretly something like `class <PkgRegistry>::Bar < PackageSpec` or something like that. How would I ever jump from an `import Bar` to the real `Bar` class?

Comment (aadi-stripe) [reply README.md:509]: Don't think we have a way currently to do that. I think that stopped working after we switched to the visibility checker design.

Comment (aadi-stripe) [reply README.md:509]: This PR improves the story a little, but does not solve that problem.

Comment (aadi-stripe) [reply README.md:509]: But we can totally fix that in a separate change

**Review by jez — APPROVED** [review]

**Pull request merged**